docs/cudf/source/developer_guide/developer_guide.md

```diff
@@ -0,0 +1,130 @@
+
+**Open question**: Should we start enforcing a stricter file/directory layout? Any suggestions?
+
+
+## Code style
+
+cuDF employs a number of linters to ensure consistent style across the code base:
+
+- [`flake8`](https://github.com/pycqa/flake8) checks for general code formatting compliance. 
+- [`black`](https://github.com/psf/black) is an automatic code formatter.
+  Code formatted with `black` is largely `flake8`-compliant, although there are a few minor issues that it cannot fix automatically.
+- [`isort`](https://pycqa.github.io/isort/) ensures imports are sorted consistently.
+- [`mypy`](http://mypy-lang.org/) performs static type checking.
+  In conjunction with [type hints](https://docs.python.org/3/library/typing.html),
+  `mypy` can help catch various bugs that are otherwise difficult to find.
+- [`pydocstyle`](https://github.com/PyCQA/pydocstyle/) lints docstring style.
+
+Configuration information for these tools is contained in a number of files.
+The primary source of truth is the `.pre-commit-config.yaml` file at the root of the repo.
+As described in 
+[overall contributing guide](https://github.com/rapidsai/cudf/blob/main/CONTRIBUTING.md),
+we recommend using [`pre-commit`](https://pre-commit.com/) to manage all linters.
+
+## Deprecating and removing code
+
+cuDF generally follows the policy of deprecating code for one release prior to removal.
+For example, if we want to remove an API after the 22.06 release, it will be marked as deprecated in the 22.08 release.
+The code may then be removed in the 22.10 release.
+All usage of deprecated APIs within cuDF should be removed when the API is deprecated.
+This prevents us from surfacing unexpected warnings to users of other APIs.
+The documentation for the API should also be updated to reflect its deprecation.
+
+When deprecating an API, developers should open a corresponding Github issue to track the API's removal.
+The GitHub issue should be labeled “deprecation” and added to the next release’s project board.
+If necessary, the removal timeline can be discussed on this issue.
+Upon removal this issue may be closed.
+Additionally, when removing an API, make sure to remove all tests and documentation.
+
+Deprecation messages should follow these guidelines:
+- Throw a FutureWarning.
+- Use a single line (no newline characters)
+- Indicate a replacement API, if any exists.
+  Deprecations are an opportunity to show users better ways to do things.
+- Don’t specify a version when removal will occur.
+  This gives us more flexibility.
+
+For example:
+```python
+warnings.warn(
+    "`Series.foo` is deprecated and will be removed in a future version of cudf. "
+    "Use `Series.new_foo` instead.",
+    FutureWarning
+)
+```
+
+```{warning}
+Deprecations should be signaled using a `FutureWarning` **not a `DeprecationWarning`**!
+`DeprecationWarning` is intended for Python's internal deprecations, not for application or library developers
```

Comment:
Can you cite where you’re pulling this from? I have seen the difference described differently. It’s not about whether the warning comes from Python’s internals, but whether the target audience is developers or application users. https://docs.python.org/3/library/warnings.html

Reply:
The best description is probably in the [original PEP that changed the recommended warnings](https://peps.python.org/pep-0565/#additional-use-case-for-futurewarning). The wording that you're referring to is more canonical (it's what shows up in the Python documentation). In practice, though, the line between developers and application users in Python is blurry enough that the only people I've seen actually recommending the use of `DeprecationWarning` in the last few years is core Python developers (see e.g. https://discuss.python.org/t/pendingdeprecationwarning-is-really-useful/1038/18). For all third-party packages `FutureWarning` is generally the safer option since you never really know how your package will be used, unless it's explicitly a command-line tool like click.

In this case I don't think this line adds much either way. I'm going to change it to focus on the important point, which is that `FutureWarning` is shown by default.